### PR TITLE
ICU-20741 Adding tests for C/C++ API into DateFormatTests

### PR DESCRIPTION
**Pull request by caiolima**
https://unicode-org.atlassian.net/browse/ICU-20741

Cc @sffc

### CODE REVIEW

File: icu4c/source/test/intltest/dtfmttst.cpp [hunk 1]
```diff
@@ -5555,6 +5556,51 @@ void DateFormatTest::TestAdoptCalendarLeak() {
...
+
+    assertEquals("Result", UnicodeString(res, length), result);
+
+    int32_t field, beginIndex, endIndex;
+    FieldPosition fieldPosition;
+    while ((field = ufieldpositer_next(fpi.getAlias(), &beginIndex, &endIndex)) >= 0) {
+        fpositer.next(fieldPosition);
+        assertEquals("Field", fieldPosition.getField(), field);
+        assertEquals("BeginIndex", fieldPosition.getBeginIndex(), beginIndex);
+        assertEquals("EndIndex", fieldPosition.getEndIndex(), endIndex);
```
Comment (sffc) [inline icu4c/source/test/intltest/dtfmttst.cpp]: Checking for C/C++ consistency is good, but you should also check the actual expected values here.  For example, have an array of the expected FieldPositions, and assert that your results match (also in length).

File: icu4c/source/test/intltest/dtfmttst.cpp [hunk 2]
```diff
@@ -5555,6 +5556,51 @@ void DateFormatTest::TestAdoptCalendarLeak() {
...
+    assertEquals("Result", result, u"W, in the afternoon");
+
+    // Now we test C API comparing result with C++
+
+    LocalUDateFormatPointer df(udat_open(UDAT_PATTERN, UDAT_PATTERN, locale.getName(),
+        timeZone, std::char_traits<char16_t>::length(timeZone),
+        pattern.getTerminatedBuffer(), pattern.length(), status));
+
+    UChar res[100];
+
```
Comment (sffc) [inline icu4c/source/test/intltest/dtfmttst.cpp]: Nit: The grouping of lines is strange.  Use double-space to separate different distinct sections, and single-space within a section.

File: icu4c/source/test/intltest/dtfmttst.cpp [hunk 3]
```diff
@@ -5555,6 +5556,51 @@ void DateFormatTest::TestAdoptCalendarLeak() {
     sdf.adoptCalendar(Calendar::createInstance(status));
 }
 
+void DateFormatTest::Test20741_ABFields() {
+    IcuTestErrorCode status(*this, "Test20741_ABFields");
+
```
Comment (sffc) [inline icu4c/source/test/intltest/dtfmttst.cpp]: Add a comment saying what this test is actually testing.  We are testing that the `a` and `B` fields are not duplicated in the field position iterator.

**Comment by sffc** [comment]
LGTM; did you have a change to check the behavior in previous ICU versions?  You could apply the diff for this test on top of the tags `release-63-1` and `release-64-2`, for example.

**Comment by srl295** [comment]
@sffc you could open an issue similar to https://github.com/cla-assistant/cla-assistant/issues/488 to get the CLA assistant people to look at it

**Comment by caiolima** [comment]
> LGTM; did you have a change to check the behavior in previous ICU versions? You could apply the diff for this test on top of the tags `release-63-1` and `release-64-2`, for example.

I'm on it now.

**Comment by caiolima** [comment]
@sffc Here is the output for 64.2:

```
-----------------------------------------------
 IntlTest (C++) Test Suite for                 
   International Components for Unicode 64.2
   Bits: 64, Byte order: Little endian, Chars: ASCII
-----------------------------------------------
 Options:                                       
   all (a)                  : Off
   Verbose (v)              : Off
   No error messages (n)    : Off
   Exhaustive (e)           : Off
   Leaks (l)                : Off
   utf-8 (u)                : Off
   notime (T)               : Off
   noknownissues (K)        : Off
   Warn on missing data (w) : Off
   Threads                  : 12
-----------------------------------------------

=== Handling test: format/DateFormatTest/Test20741_ABFields: ===
   format {
      DateFormatTest {
         Test20741_ABFields {
         FAIL: Result; got "W, in the afternoon"; expected "W, p"
         FAIL: Field in C++; got 66=0x42; expected 97=0x61
         FAIL: EndIndex in C++; got 19=0x13; expected 4=0x04
         FAIL: Field in C++; got 0=0x00; expected 66=0x42
         FAIL: BeginIndex in C++; got 0=0x00; expected 3=0x03
         FAIL: EndIndex in C++; got 5439568=0x530050; expected 4=0x04
         FAIL: Fields count; got 2=0x02; expected 3=0x03
         FAIL: Result; got "W, in the afternoon"; expected "W, p"
         FAIL: Field in C API; got 66=0x42; expected 97=0x61
         FAIL: EndIndex in C API; got 19=0x13; expected 4=0x04
         FAIL: Field in C API; got 0=0x00; expected 66=0x42
         FAIL: BeginIndex in C API; got 0=0x00; expected 3=0x03
         FAIL: EndIndex in C API; got 5439568=0x530050; expected 4=0x04
         FAIL: Fields count; got 2=0x02; expected 3=0x03
      
         } ERRORS (14) in Test20741_ABFields
      
   
      } ERRORS (14) in DateFormatTest
   

   } ERRORS (14) in format


--------------------------------------
Errors in total: 14.
            Test20741_ABFields
         DateFormatTest
      format
   
--------------------------------------
```

I was not able to run it on 63.2, since it breaks during `make check`.

```
DYLD_LIBRARY_PATH=../../lib:../../stubdata:../../tools/ctestfw:$DYLD_LIBRARY_PATH  ./../../bin/uconv -f ibm-37 ../../../icu/icu4c/source/extra/uconv/samples/ibm-37-test.txt
./../../bin/uconv: can not initialize ICU.  status = U_FILE_ACCESS_ERROR
make[2]: *** [check-local] Error 1
make[1]: *** [check-recursive] Error 2
make: *** [check-recursive] Error 2
```

**Comment by sffc** [comment]
Great; thanks for that.  It looks like the issue was present in ICU 64 and no longer ICU 65.  This test therefore prevents this from regressing.

One thing I noticed from the test output.  The convention for assertEquals is,

    assertEquals(message, expected, actual)

It looks like your code currently mixes up expected and actual.  Please fix the order, which will also fix the error messages.

**Comment by mrego** [comment]
@caiolima is now part of `igalia-google-contributors` group, is that good enough regarding CLA issues here?

**Comment by sffc** [comment]
Looks good, but also please see @anba's comment on the ICU issue thread.  This issue may still exist in other locales, just not en.

Please:

1. Change the test to test more locales
2. If necessary, use quick mode to speed up the test: `if (quick) { ... }`, for example, to skip locales.  Usually you can skip about 90% of locales in quick mode.  The full test can be run in exhaustive mode by passing `-e` to intltest, which also runs automatically in continuous integration.
3. If necessary, use `logKnownIssue` to protect failing tests.  Search for examples.

**Comment by sffc** [comment]
Ping

**Comment by caiolima** [comment]
@sffc I updated test to verify is there is any locale with 2 fields being present to the same substring. I'm not sure if `if (quick)` is necessary there, since test doesn't take long to finish. I also removed the C test, since C++ API is reporting same errors.

File: icu4c/source/test/intltest/dtfmttst.cpp [hunk 4]
```diff
@@ -5555,6 +5556,44 @@ void DateFormatTest::TestAdoptCalendarLeak() {
...
+ */
+void DateFormatTest::Test20741_ABFields() {
+    IcuTestErrorCode status(*this, "Test20741_ABFields");
+
+    const char16_t timeZone[] = u"PST8PDT";
+
+    UnicodeString skeleton = u"EEEEEBBBBB"; 
+    int32_t count = 0;
+    const Locale* locales = Locale::getAvailableLocales(count);
+    for (int32_t i = 0; i < count; i++) {
```
Comment (sffc) [inline icu4c/source/test/intltest/dtfmttst.cpp]: Looping over all locales is too slow.  I just tested it.  We try to keep tests at no more than 100 ms (hopefully faster), and this takes 910 ms.

Please add:

```cpp
if (quick && (i % 17) != 0) { continue; }
```

to make the test run on 1/17th of locales, except in exhaustive mode when it still runs on all locales.

**Comment by sffc** [comment]
For Java: this is one of the last APIs that hasn't been upgraded to FormattedValue for easier iteration over field positions.  Sorry about that.  [ICU-20897](https://unicode-org.atlassian.net/browse/ICU-20897) might fix that, but for now we have to use AttributedCharacterIterator.

The gist of it is:

```java
AttributedCharacterIterator aci = // get your AttributedCharacterIterator
for (char c = fpi.first(); c != AttributedCharacterIterator.DONE; c = fpi.next(), i++) {
    Set<AttributedCharacterIterator.Attribute> currentAttributes = fpi.getAttributes().keySet();
    // currentAttributes is a set of attributes for the current *character*
}
```

In other words, you get attributes on a character-by-character basis, unlike field spans as in C++.

**Comment by sffc** [comment]
Hi Caio, any update on this?  I'd like to get the tests checked in so you can start working on the fix :)

**Comment by caiolima** [comment]
> Hi Caio, any update on this? I'd like to get the tests checked in so you can start working on the fix :)

Hi Shane. Did you see that I uploaded a version of the Java test using a different API? I was still able to reproduce failures for the same locales there.

**Comment by sffc** [comment]
> > Hi Caio, any update on this? I'd like to get the tests checked in so you can start working on the fix :)
> 
> Hi Shane. Did you see that I uploaded a version of the Java test using a different API? I was still able to reproduce failures for the same locales there.

OK! Looks good; there may have been a race condition because I didn't see any more posts since my https://github.com/unicode-org/icu/pull/875#issuecomment-554618536, which was in response to your question over email.  I think this looks good to check in.  Now you can start working on the actual fix. :)

**Pull request merged**